### PR TITLE
keep JSyntaxTextArea text value for use in headless mode

### PR DESCRIPTION
**Pull request by stokpop**
## Description
Keep the text that is set on overridden JSyntaxTextArea anonymous class in headless mode.

## Motivation and Context
Working on a project that uses jMeter components in a headless way (`java.awt.headless=true`), we found that the `setText(...)` is not kept on the `JSyntaxTextArea` from `getInstance(...)`. Following `getText()` call returns null. With `java.awt.headless=false`, to try and run with the original `JSyntaxTextArea` on headless linux it fails as there is no X11 display. 

This might benefit headless unit test as well.

## How Has This Been Tested?
Unit tests added. Snapshot build works in project.

## Types of changes
- New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code follows the [code style][style-guide] of this project.
- [x] I have updated the documentation accordingly (javadocs and code comments)

### CODE REVIEW

File: src/core/src/main/java/org/apache/jmeter/gui/util/JSyntaxTextArea.java [hunk 1]
```diff
@@ -118,7 +119,15 @@ public void setCaretPosition(int b) { }
                     @Override
                     public void discardAllEdits() { }
                     @Override
-                    public void setText(String t) { }
+                    public void setText(String t) {
```
Comment (vlsi) [inline src/core/src/main/java/org/apache/jmeter/gui/util/JSyntaxTextArea.java:122]: Just wondering, does the removal of `setText` override solves the issue?

Comment (FSchumacher) [reply src/core/src/main/java/org/apache/jmeter/gui/util/JSyntaxTextArea.java:122]: If I add a `super.setText(t)` here, I get the following stacktrace:
```
java.lang.NullPointerException
        at org.fife.ui.rsyntaxtextarea.RSyntaxTextArea.getFontMetricsForTokenType(RSyntaxTextArea.java:1204)
        at org.fife.ui.rsyntaxtextarea.TokenImpl.getWidthUpTo(TokenImpl.java:578)
        at org.fife.ui.rsyntaxtextarea.TokenImpl.getWidth(TokenImpl.java:570)
        at org.fife.ui.rsyntaxtextarea.WrappedSyntaxView.calculateBreakPosition(WrappedSyntaxView.java:127)
        at org.fife.ui.rsyntaxtextarea.WrappedSyntaxView$WrappedLine.calculateLineCount(WrappedSyntaxView.java:1191)
        at org.fife.ui.rsyntaxtextarea.WrappedSyntaxView$WrappedLine.handleDocumentEvent(WrappedSyntaxView.java:1440)
        at org.fife.ui.rsyntaxtextarea.WrappedSyntaxView$WrappedLine.insertUpdate(WrappedSyntaxView.java:1464)
        at org.fife.ui.rsyntaxtextarea.WrappedSyntaxView.insertUpdate(WrappedSyntaxView.java:705)
        at javax.swing.plaf.basic.BasicTextUI$RootView.insertUpdate(BasicTextUI.java:1610)
        at javax.swing.plaf.basic.BasicTextUI$UpdateHandler.insertUpdate(BasicTextUI.java:1869)
        at javax.swing.text.AbstractDocument.fireInsertUpdate(AbstractDocument.java:201)
        at org.fife.ui.rsyntaxtextarea.RSyntaxDocument.fireInsertUpdate(RSyntaxDocument.java:187)
        at javax.swing.text.AbstractDocument.handleInsertString(AbstractDocument.java:748)
        at javax.swing.text.AbstractDocument.insertString(AbstractDocument.java:707)
        at javax.swing.text.PlainDocument.insertString(PlainDocument.java:130)
        at javax.swing.text.AbstractDocument.replace(AbstractDocument.java:669)
        at javax.swing.text.JTextComponent.setText(JTextComponent.java:1669)
        at org.apache.jmeter.gui.util.JSyntaxTextArea$1.setText(JSyntaxTextArea.java:122)
```
So, I would say, no, it does not resolve the issue :)

File: src/core/src/main/java/org/apache/jmeter/gui/action/Save.java [hunk 2]
```diff
@@ -92,7 +92,7 @@
     private static final int BACKUP_MAX_COUNT = JMeterUtils.getPropDefault(JMX_BACKUP_MAX_COUNT, 10);
 
     // NumberFormat to format version number in backup file names
-    private static final DecimalFormat BACKUP_VERSION_FORMATER = new DecimalFormat("000000"); //$NON-NLS-1$
+    private static final DecimalFormat BACKUP_VERSION_FORMAT = new DecimalFormat("000000"); //$NON-NLS-1$
```
Comment (FSchumacher) [inline src/core/src/main/java/org/apache/jmeter/gui/action/Save.java:95]: This has nothing to do with the description of the issue. I don't think it is needed here and should be done in it's own PR (if at all).

File: src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java [hunk 3]
```diff
@@ -41,7 +45,37 @@ public void testSetLanguage() {
             textArea.setLanguage(null);
             assertEquals(SyntaxConstants.SYNTAX_STYLE_NONE, textArea.getSyntaxEditingStyle());
         } catch (HeadlessException he) {
-            // Does not work in headless mode
+            // Does not work in headless mode, which depends on value of java.awt.headless property
+            // and the OS (e.g. might work on MacOS and not on Linux due to missing X11).
```
Comment (FSchumacher) [inline src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java:49]: I don't understand this new comment. Can you explain it to me?

File: src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java [hunk 4]
```diff
@@ -41,7 +45,37 @@ public void testSetLanguage() {
...
+            System.out.println("WARNING for JSyntaxTextAreaTest.testSetLanguage test: does not work in headless mode");
+        }
+    }
+
+    @Test
+    public void testHeadless() {
+        String key = "java.awt.headless";
+        String initialValue = System.getProperty(key);
+        try {
+            System.setProperty(key, "true");
```
Comment (FSchumacher) [inline src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java:59]: Is this really a sane way to test headless mode?

File: src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java [hunk 5]
```diff
@@ -41,7 +45,37 @@ public void testSetLanguage() {
...
+            // getInstance returns anonymous class with some overridden methods
+            // to avoid errors due to 'java.awt.headless=true'
+            JSyntaxTextArea textArea = JSyntaxTextArea.getInstance(10,20);
+
+            String myText = "my text";
+            textArea.setText(myText);
```
Comment (FSchumacher) [inline src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java:65]: Are you testing set/getText here, or the creation of a specialised subclass?

File: src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java [hunk 6]
```diff
@@ -41,7 +45,37 @@ public void testSetLanguage() {
...
+            // next one fails with NPE when getInstance does not provide overridden method
+            textArea.setCodeFoldingEnabled(true);
+            assertEquals(myText, textArea.getText());
+
+        } catch (HeadlessException he) {
+            fail("WARNING: Does not work in headless mode");
```
Comment (FSchumacher) [inline src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java:71]: Is this expected? I would expect, that this fail should never be reached and should point this out a bit more clearly.

File: src/core/src/main/java/org/apache/jmeter/gui/action/Save.java (hunk 2 above)
Comment (stokpop) [reply src/core/src/main/java/org/apache/jmeter/gui/action/Save.java]: true, typo unrelated to PR, removed it

File: src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java (hunk 5 above)
Comment (stokpop) [reply src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java:65]: Test if the textArea returned by getInstance() in headless mode is retaining the text. Without the fix of the specialised subclass this will fail, because getText() returns null.

Comment (FSchumacher) [reply src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java:65]: Right, but why the call ` textArea.setCodeFoldingEnabled(true);`? It seems that you want to test, that we get a special variant, that works in headless mode. But in my eyes, you know, that it works in headless mode (or could not check it anyway) and this test is about saving the text. One test should assert one thing (at least new ones should :))

File: src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java (hunk 3 above)
Comment (stokpop) [reply src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java:49]: How I got here: I have Kotlin code (to generate a jmx file based on open-api spec) that calls setText() and getText() for the json payload. When running from Kotlin code on MacOS laptop, the code runs fine in Intellij: getText() returns a value. When running it from Kotlin _script_, the getText() returns null after setText(). Why? Debugging shows that `java.awt.headless=true` when running Kotlin _script_. When running Kotlin code the `java.awt.headless` property is not present, so does not run headless. (In that case it I guess it might work because `awt.toolkit=sun.lwawt.macosx.LWCToolkit` is present). Then on linux docker (target system) running non-headless does not work because it says no DISPLAY is set (awt.toolkit is missing?). And running headless (with `java.awt.headless=true`), the getText() returns null.

File: src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java (hunk 4 above)
Comment (stokpop) [reply src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java:59]: Maybe there is a better way? It is indeed tricky as it might impact other tests, that's why the `@Isolated` is added. The code directly depends on the property value in its behaviour. I've seen some mock framework code that overrides the System.getProperty() to return a certain value, but there might be classloading issues.

File: src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java (hunk 6 above)
Comment (stokpop) [reply src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java:71]: The exception is not expected here, because the explicit `java.awt.headless=true`, so the fail. This is indeed different from the other test, that seems to allow the HeadlessException by ignoring it. I can change fail text.

File: src/core/src/main/java/org/apache/jmeter/gui/util/JSyntaxTextArea.java (hunk 1 above)
Comment (stokpop) [reply src/core/src/main/java/org/apache/jmeter/gui/util/JSyntaxTextArea.java:122]: True, I saw same NPE.

File: src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java (hunk 5 above)
Comment (stokpop) [reply src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java:65]: agree, I removed the NPE implicit check

File: src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java (hunk 3 above)
Comment (vlsi) [reply src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java:49]: I've filed https://github.com/bobbylight/RSyntaxTextArea/issues/424 to `RSyntaxTextArea`. I hope we can remove headless workaround then.

File: src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java (hunk 4 above)
Comment (FSchumacher) [reply src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java:59]: For the other tests, we are expecting the test suite to be set up in the correct way (for example most of our CI stuff is done in a headless setup). But if this works, we can use it.

File: src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java [hunk 7]
```diff
@@ -41,7 +45,35 @@ public void testSetLanguage() {
...
+            System.setProperty(key, "true");
+            // getInstance returns anonymous class with some overridden methods
+            // to avoid errors due to 'java.awt.headless=true'
+            JSyntaxTextArea textArea = JSyntaxTextArea.getInstance(10,20);
+
+            String myText = "my text";
+            textArea.setText(myText);
+            assertEquals(myText, textArea.getText());
+
+        } catch (HeadlessException he) {
```
Comment (FSchumacher) [inline src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java:68]: I think we could drop this catch, as every exception here is unexpected and should be seen after.

Comment (stokpop) [reply src/core/src/test/java/org/apache/jmeter/gui/util/JSyntaxTextAreaTest.java]: Agree, I dropped the catch.
